[Jiacheng] iP

Jiacheng-y opened the pull request:
### Duke

> "Your mind is for having ideas, not holding them." -David Allen ([source](https://dansilvestre.com/productivity-quotes))

This new and upgraded version of Duke is 

- text-based
- easy to learn
- ~really~ _super_ fast to use

All you need to do is 

1. download it from [here](https://github.com/Jiacheng-y/ip/releases)
3. open terminal
4. navigate to the folder that contains the downloaded .jar file on your computer
5. run the command `java -jar duke.jar`
6. add your tasks
7. Let it manage your tasks! 🤩

Best of all, it is **FREE**! 

Features: 

- [ ] Managing tasks
- [ ] Managing deadlines
- [ ] Graphic Interface (coming soon)

Here's the `main` method: 

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Jiacheng-y committed:
level 1 task: greet, echo, exit

Jiacheng-y committed:
Level-2: add, list

Jiacheng-y committed:
Level-3: Mark as done is done

Jiacheng-y committed:
Level-4: ToDos, Events, deadlines

Jiacheng-y committed:
ui-text-testing : added test cases

Jiacheng-y committed:
Level-5: handling errors

Jiacheng-y committed:
Level-5: handling errors

Jiacheng-y committed:
Level-6 delete tasks

Jiacheng-y committed:
Level-6 delete tasks

Jiacheng-y committed:
adds increment for Level-7: saves tasks to txt file

Jiacheng-y committed:
Add function to save tasks to txt file for Level-7

Jiacheng-y committed:
adds date class for Level-8

Jiacheng-y committed:
Merge branch 'branch-Level-7'
Add Level-7 increments to master: allow tasks to be saved to txt file
and read from txt file

Jiacheng-y committed:
Merge branch 'branch-Level-8'
DateTime Class: recognise date inputs as dates instead of Strings

Jiacheng-y committed:
Categorise objects and functionalities into different classes

Jiacheng-y committed:
organise classes into packages for A-Packages: Exception, Task, Tools

Jiacheng-y committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

Jiacheng-y committed:
add gradle for A-Gradle

Jiacheng-y committed:
Add gradle for A-Gradle

Jiacheng-y committed:
Add tests for CommandParser and DateTimeParser for A-JUnit

Jiacheng-y committed:
add tests for CommandParser and DateTimeParser for A-JUnit

Jiacheng-y committed:
package app as JAR file for A-Jar

Jiacheng-y committed:
Package app as JAR file for A-Jar

Jiacheng-y committed:
add JavaDocs to some classes for A-JavaDocs

Jiacheng-y committed:
Add JavaDocs to classes for A-JavaDocs

Jiacheng-y committed:
amend code based on coding style for A-CodingStyle, and integrate stylechecks for A-Stylecheck

Jiacheng-y committed:
Amend code based on coding style for A-CodingStyle

Jiacheng-y committed:
Amend code based on coding style for A-CodingStyle

Jiacheng-y committed:
amend code based on coding style for A-CodingStyle, and integrate stylechecks for A-Stylecheck

Jiacheng-y committed:
add search function to find tasks containing keyword.

Jiacheng-y committed:
Add search function to find tasks with keyword.

Jiacheng-y committed:
Merge branch 'branch-Level-9'
add search word among tasks to Duke

Jiacheng-y committed:
edit code to fit coding standards

Jiacheng-y committed:
add JavaDocs to the project

Jiacheng-y committed:
Change directory that tasks.txt is created in

Jiacheng-y committed:
Change directory that tasks.txt is created in

weesiau commented on .gitignore:
```diff
@@ -15,3 +16,5 @@ bin/
 
 /text-ui-test/ACTUAL.txt
 text-ui-test/EXPECTED-UNIX.TXT
+*.class
+*.jar
```
I like the inclusion of these files in the gitignore

weesiau commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,103 @@
+                    try {
+                        tasks.addTask(input);
+                        storage.saveTasks(tasks);
+                    } catch (TaskNotFoundException e) {
+                        Ui.noTaskExceptionToast(e);
+                    } catch (ContentNotFoundException e) {
+                        Ui.noContentExceptionToast(e);
+                    } catch (DateNotFoundException e) {
+                        Ui.noDateExceptionToast(e);
+                    } catch (DateTimeParseException e) {
+                        Ui.wrongDateFormatToast(e);
```
Great effort catching all exceptions individually

weesiau commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,45 @@
+package duke.task;
```
Good use of detailed packages, applies to all your files

weesiau commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,58 @@
+ * Task in the task management.
+ */
+public class Task {
+    private String task;
+    private boolean done;
```
Could these attributes be named more descriptively?

weesiau commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,58 @@
+     * Marks a task as completed.
+     * @return The new completion status of the task: true.
+     */
+    public boolean done() {
+        done = true;
+        return true;
+    }
+
+    /**
+     * Marks a task as not completed.
+     * @return The new completion status of the task: false.
+     */
+    public boolean notDone() {
```
Similarly, can these methods be named more descriptively?

weesiau commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,130 @@
+        Ui.printList(tasks);
+    }
+
+    public Task markTask(Integer n) {
```
Perhaps the parameter can be named more descriptively?

weesiau reviewed (commented):
Overall, your code is well written and well organised. I like the different tools and tricks you used, but some variable names can be more descriptive.

Jiacheng-y committed:
Add GUI to duke

Jiacheng-y committed:
Add GUI to duke

Jiacheng-y committed:
Edit GUI implementation

Jiacheng-y committed:
Edit GUI implementation

Jiacheng-y committed:
Edit GUI to interact with user

Jiacheng-y committed:
Edit code for coding standard

Jiacheng-y committed:
Clean up GUI related classes

Jiacheng-y committed:
Merge branch 'branch-Level-10'

Jiacheng-y committed:
Merge branch 'master' into branch-Level-10
Merge master branch into branch-Level-10

Jiacheng-y committed:
Add JavaDocs to GUI classes

Jiacheng-y committed:
Merge branch 'branch-Level-10'
Add JavaDocs to GUI classes

Jiacheng-y committed:
Add assertions and fix GUI bugs

1. No use of assertions in code.

Adding assertions can document important assumptions hold in relevant
methods.

Let's add assertions to define runtime assumptions about the program
state. For example, assertiosn can be used to define storage cannot be
null during runtime.

2. Bugs present in code
* Fix NullPointerException caught by assert statement

Jiacheng-y committed:
Merge PR#2 from Jiacheng-y/branch-A-Assertions

Add assertions and fix GUI bugs

1. No use of assertions in code.

Adding assertions can document important assumptions hold in relevant
methods.

Let's add assertions to define runtime assumptions about the program
state. For example, assertiosn can be used to define storage cannot be
null during runtime.

2. Bugs present in code
* Fix NullPointerException caught by assert statement

Jiacheng-y committed:
Refactor code to improve code quality

Some regions of code are lengthy and not well organised, making the code
difficult to read and debug.

Methods and classes are refactord to improve readability.

Let's
- shorted method lengths.
- make the prominent path obvious.
- reduce arrow head shaped methods.

Jiacheng-y committed:
Merge pull request #3 from Jiacheng-y/branch-A-CodeQuality

Refactor code to improve code quality

Some regions of code are lengthy and not well organised, making the code
difficult to read and debug.

Methods and classes are refactord to improve readability.

Let's
- shorted method lengths. 
- make the prominent path obvious. 
- reduce arrow head shaped methods.

Jiacheng-y committed:
Merge branch-Level-7 to master

Jiacheng-y committed:
Add sort functionality

Tasks were listed in order of being added.

Add comparators to sort according to :
- due date of task
- name of task

Jiacheng-y committed:
Make minor stylistic edits

Jiacheng-y committed:
Merge pull request #4 from Jiacheng-y/branch-C-Sort

Merge branch-C-Sort to master

Add sorting functionality to Duke.

Tasks can only be sorted based on order of adding the tasks.

Add sorting based on :

Due date
Task description
Task type
Completion status

Jiacheng-y committed:
Modify dialogue box style

Jiacheng-y committed:
Edit some stylistic changes

Jiacheng-y committed:
Add demonstration img

Jiacheng-y committed:
Update README.md

Jiacheng-y committed:
Troubleshoot GUI file paths

Jiacheng-y committed:
Merge branch 'master' of github.com/Jiacheng-y/ip

Jiacheng-y committed:
Modify README.md

Jiacheng-y committed:
Update README.md

Jiacheng-y committed:
Handle more exceptions

Exceptions written to handle more edge cases:
* Array out of bounds error for mark, unmark, and delete functions
* Missing keywords for find, sort, todo, deadline, event

Jiacheng-y committed:
Merge pull request #5 from Jiacheng-y/branch-MoreExceptions

Handle more edge cases
